File: java/patterns/src/main/java/org/prototypepatterns/structural/facade/FacadePattern.java
package org.prototypepatterns.structural.facade;

public final class FacadePattern {
    private FacadePattern() {
    }

    static final class Lights {
        String dim() {
            return "lights dimmed";
        }
    }

    static final class Projector {
        String on() {
            return "projector on";
        }
    }

    static final class SoundSystem {
        String surround() {
            return "surround enabled";
        }
    }

    static final class HomeTheaterFacade {
        private final Lights lights;
        private final Projector projector;
        private final SoundSystem soundSystem;

        HomeTheaterFacade(Lights lights, Projector projector, SoundSystem soundSystem) {
            this.lights = lights;
            this.projector = projector;
            this.soundSystem = soundSystem;
        }

        String watchMovie() {
            return lights.dim() + " | " + projector.on() + " | " + soundSystem.surround();
        }
    }

    public static String demo() {
        return new HomeTheaterFacade(new Lights(), new Projector(), new SoundSystem()).watchMovie();
    }
}

File: csharp/patterns/src/PrototypePatterns/Structural/Facade/FacadePattern.cs
namespace PrototypePatterns.Structural.Facade;

public sealed class Lights
{
    public string Dim() => "lights dimmed";
}

public sealed class Projector
{
    public string On() => "projector on";
}

public sealed class SoundSystem
{
    public string Surround() => "surround enabled";
}

public sealed class HomeTheaterFacade(Lights lights, Projector projector, SoundSystem soundSystem)
{
    public string WatchMovie() => $"{lights.Dim()} | {projector.On()} | {soundSystem.Surround()}";
}

public static class FacadePattern
{
    public static string Demo() => new HomeTheaterFacade(new Lights(), new Projector(), new SoundSystem()).WatchMovie();
}
